added new section for  Configuring how many processes can be launched… (#1970)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc

[id="configuring_how_many_processes_can_be_launched_by_apache_httpd_{context}"]
= Configuring how many processes can be launched by Apache httpd

.Configuring how many processes can be launched by Apache httpd

By default, HTTPD uses event request handling mechanism.

When the number of requests to the apache exceed the maximum number of child processes that can be launched to handle the incoming connections, an HTTP 503 Service Unavailable Error is raised by httpd.
Amidst httpd running out of processes to handle, the incoming connections can also result in multiple component failure on your Foreman side due to the dependency of components like Pulp on the availability of httpd processes.

You can adapt the configuration of HTTPD event to handle more concurrent requests based on your expected peak load.

> . You can modify these values in /etc/foreman-installer/custom-hiera.yaml by changing or adding these lines:

Reviewer: Hello Imaan. We have a generic section that talks about how to perform the configuration change. Besides need for installer, it also mentions backups and sanity checks after the procedure. Please take a look at other places of tuning guide where we recommend some config change and use same stuff.

> . You can modify these values in /etc/foreman-installer/custom-hiera.yaml by changing or adding these lines:

Author: Updated. Thank you @jhutar & @maximiliankolb

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
apache::mod::event::serverlimit: 64
apache::mod::event::maxrequestworkers: 1024
apache::mod::event::maxrequestsperchild: 4000
----
+

> These are the numbers installer will apply when using option --tuning=medium.

Reviewer 2: I'm going to say that this needs a big disclaimer that it breaks the tuning functionality. The above values are what the installer sets with `--tuning medium` (https://github.com/theforeman/foreman-installer/blob/develop/config/foreman.hiera/tuning/sizes/medium.yaml) and above. If users place these values in `custom-hiera.yaml` we can never override them again.

Author: Hello @ekohl : I can understand your point and totally agree on this but we got a request from customers to update this section soon as mentioned in this bug - https://bugzilla.redhat.com/show_bug.cgi?id=2158458

As performance testing is still in progress (we cannot propose any number as of now), so meanwhile we are showing `how to do the change and what is the name of the variable` . Let me add a waring message in this section. Please feel free to change or suggest us. Thank you in advance.

[id="configuring_how_many_processes_can_be_launched_by_apache_httpd_{context}"]
= Configuring how many processes can be launched by Apache httpd

By default, HTTPD uses event request handling mechanism.

> When the number of requests to httpd exceed the maximum number of child processes that can be launched to handle the incoming connections, httpd raises an HTTP 503 Service Unavailable error.

Reviewer 3: ```suggestion
When the number of requests to httpd exceeds the maximum number of child processes that can be launched to handle the incoming connections, httpd raises an HTTP 503 Service Unavailable error.
```

> Amidst httpd running out of processes to handle, the incoming connections can also result in multiple component failure on your {Project} services side due to the dependency of components like Pulp on the availability of httpd processes.

Reviewer 3: ```suggestion
Amidst httpd running out of processes to handle, the incoming connections can also result in multiple component failures on your {Project} services side due to the dependency of components like Pulp on the availability of httpd processes.
```

You can adapt the configuration of HTTPD event to handle more concurrent requests based on your expected peak load.

.Procedure
. Modify the number of concurrent requests in '/etc/foreman-installer/custom-hiera.yaml' by changing or adding the following lines:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
apache::mod::event::serverlimit: 64
apache::mod::event::maxrequestworkers: 1024
apache::mod::event::maxrequestsperchild: 4000
----

> +

Reviewer 4: ```suggestion
```

. Apply your changes to {ProjectServer}.
For more information, see xref:Applying_configurations_{context}[].

> These are the numbers installer will apply when using option --tuning=medium.

Reviewer 4: ```suggestion
The example is identical to running `{foreman-installer} --tuning=medium` on {ProjectServer}.
```

Unsure about the wording. Any alternatives @Lennonka ?

Reviewer: Let's add a warning like (as mentioned by Evgeni): `... but be warned that configuring these numbers in custom-hiera.yaml will effectively lock them, so if {Product} installer decides to change these numbers in the future by --tuning=* parameter, the possible change will be overwritten by this custom-hiera.yaml setting. Please use your numbers only if you have specific need for it.`

Author: Thank you @jhutar . Added a note.

[id="configuring_how_many_processes_can_be_launched_by_apache_httpd_{context}"]

> = Configuring how many processes can be launched by Apache httpd

Reviewer 5: Could we rephrase the title to something like "Tuning Apache httpd child processes"?
Note that the ID and the filename need also to be changed.

Author: Done. Thanks for the suggestion.

By default, HTTPD uses event request handling mechanism.
When the number of requests to httpd exceeds the maximum number of child processes that can be launched to handle the incoming connections, httpd raises an HTTP 503 Service Unavailable error.
Amidst httpd running out of processes to handle, the incoming connections can also result in multiple component failures on your {Project} services side due to the dependency of components like Pulp on the availability of httpd processes.

You can adapt the configuration of HTTPD event to handle more concurrent requests based on your expected peak load.

.Procedure
. Modify the number of concurrent requests in '/etc/foreman-installer/custom-hiera.yaml' by changing or adding the following lines:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
apache::mod::event::serverlimit: 64
apache::mod::event::maxrequestworkers: 1024
apache::mod::event::maxrequestsperchild: 4000
----
. Apply your changes to {ProjectServer}.
For more information, see xref:Applying_configurations_{context}[].

The example is identical to running `{foreman-installer} --tuning=medium` on {ProjectServer}.

> .Note
> . But be warned that configuring these numbers in custom-hiera.yaml will effectively lock them, so if {Product} installer decides to change these numbers in the future by --tuning=* parameter, the possible change will be overwritten by this custom-hiera.yaml setting. Please use your numbers only if you have specific need for it.

Reviewer 4: ```suggestion
[NOTE]
====
But be warned that configuring these numbers in custom-hiera.yaml will effectively lock them, so if {Product} installer decides to change these numbers in the future by --tuning=* parameter, the possible change will be overwritten by this custom-hiera.yaml setting. Please use your numbers only if you have specific need for it.
====
```

two more things: please reword "be warned" and optionally choose a different admonition such as `CAUTION`.

Reviewer 5: Perhaps a WARNING and cutting the initial sentence to "Configuring these numbers..."?

Author: Updated. Thanks :)
